packages/ui/src/components/controls/NumberInput.tsx

```diff
@@ -0,0 +1,126 @@
+import clsx from 'clsx';
+import type {JSX} from 'preact';
+import {useCallback, useEffect, useRef, useState} from 'preact/hooks';
+import {useDocumentEvent} from '../../hooks';
+import {MouseButton, clamp} from '../../utils';
+import styles from './Controls.module.scss';
+
+type NumberInputProps = Omit<
+  JSX.HTMLAttributes<HTMLInputElement>,
+  'min' | 'max' | 'step' | 'value' | 'onChange' | 'onChangeCapture'
+> & {
+  min?: number;
+  max?: number;
+  step?: number;
+  value: number;
+  onChange?: (value: number) => void;
+  onChangeCapture?: (value: number) => void;
+};
+
+export function NumberInput({
+  min = -Infinity,
+  max = Infinity,
+  step = 1,
+  value = 0,
+  onChange,
+  onChangeCapture,
+  ...props
+}: NumberInputProps) {
+  const handleChange = onChangeCapture ? onChange : undefined;
+  const handleChangeCapture = onChangeCapture ?? onChange;
+
+  const inputRef = useRef<HTMLInputElement>();
+  const [currentValue, setCurrentValue] = useState(value);
+  const [active, setActive] = useState(false);
```

Comment:
This state is not necessary since the value is only ever retrieved in event handles.
So whether the field has focus can be checked with just `document.activeElement === inputRef.value`.
Similarly, the styles can be changed using the `:focus` pseudo-class.